lib/classes/organisations_user_builder.rb

```diff
@@ -11,91 +13,149 @@ class OrganisationsUserBuilder
   #     :email
   #   }
 
-  def initialize(params, app_name)
-    @organisations_user = OrganisationsUser.find_by_id(params["id"]) if params["id"]
-    @user = @organisations_user.user if @organisations_user
-    @organisation_id = params["organisation_id"].presence.try(:to_i)
-    @user_attributes = params["user_attributes"]
-    @mobile = @user_attributes["mobile"].presence.try(:to_s)
-    @email = @user_attributes["email"].presence
-    @position = params["position"]
-    @preferred_contact_number = params["preferred_contact_number"]
-    @app_name = app_name
-    fail_with_error(I18n.t("organisations_user_builder.organisation.blank")) unless @organisation_id
-    fail_with_error(I18n.t("organisations_user_builder.user.mobile.blank")) unless @mobile
+  # ------------------------
+  # Entry points
+  # ------------------------
+
+  def self.create(organisations_user_params)
+    OrganisationsUserBuilder.new(organisations_user_params.symbolize_keys).create!
+  end
+
+  def self.update(organisations_user_id, organisations_user_params)
+    OrganisationsUserBuilder.new(organisations_user_params.symbolize_keys).update!(organisations_user_id)
   end
 
-  def build
-    users = User.where("lower(email) = (?) OR mobile = (?)", @email&.downcase, @mobile)
-    return fail_with_error(I18n.t('organisations_user_builder.invalid.user')) if users.count > 1
+  # ------------------------
+  # Implementation
+  # ------------------------
+
+  def initialize(organisation_id: nil, user_id: nil, user_attributes: nil, position: '', preferred_contact_number: '', status: '', change_author: User.current_user)
+    @change_author            = change_author
+    @user_id                  = user_id.to_i
+    @organisation_id          = organisation_id.to_i
+    @position                 = position
+    @status                   = status
+    @preferred_contact_number = preferred_contact_number
+    @user_attributes          = user_attributes&.symbolize_keys
+    @user                     = strict_find!(User, user_id)
+    @organisation             = strict_find!(Organisation, organisation_id)
+  end
+
+  def create!
+    @status = OrganisationsUser::Status::PENDING if @status.blank?
 
-    @user = build_user(users)
-    return fail_with_error(@user.errors) unless @user.valid?
-    return fail_with_error(I18n.t("organisations_user_builder.organisation.not_found")) unless organisation
+    assert_non_existing!
+    assert_permissions!
```

Comment:
Just a thought:
Would it better to do an early return even before creating an instance for this permission check from [here](https://github.com/crossroads/api.goodcity/pull/1080/files#diff-27eae9ac3f37050f9c9f0128aff14289R21)

Reply:
It's a good point. but some checks require records to actually be loaded first, for example checking the role.
So if you want the check in the static method, that would actually move logic out of the instanciated class, and would split things in a weird way.

I originally wanted to to use an instance at all, but there were just so many different params and variables that I caved :(